handlers/playmobile/playmobile.go

```diff
@@ -0,0 +1,191 @@
+//         <content type="text/plain">SMS Response</content>
+//     </message>
+// </sms-request>
+
+type mtResponse struct {
+	XMLName xml.Name `xml:"sms-request"`
+	Message []struct {
+		ID         string `xml:"id,attr"`
+		MSIDSN     string `xml:"msisdn,attr"`
+		SubmitDate string `xml:"submit-date,attr"`
+		Content struct {
+			Text string `xml:",chardata"`
+		} `xml:"content"`
+	} `xml:"message"`
+}
+
+// receiveMessage is our HTTP handler function for incoming messages
+func (h *handler) receiveMessage(ctx context.Context, c courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	payload := &mtResponse{}
+	err := handlers.DecodeAndValidateXML(payload, r)
+
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, c, w, r, err)
+	}
+
+	if len(payload.Message) == 0 {
+		return nil, handlers.WriteAndLogRequestIgnored(ctx, h, c, w, r, "no messages, ignored")
+	}
+
+	msgs := make([]courier.Msg, 0, 1)
+
+	// parse each inbound message
+	for _, pmMsg := range payload.Message {
+		if pmMsg.MSIDSN == "" || pmMsg.ID == "" {
+			return nil, handlers.WriteAndLogRequestError(ctx, h, c, w, r, fmt.Errorf("missing required fields msidsn or id"))
+		}
+
+		// create our URN
+		urn, err := handlers.StrictTelForCountry(pmMsg.MSIDSN, c.Country())
+		if err != nil {
+			return nil, handlers.WriteAndLogRequestError(ctx, h, c, w, r, err)
+		}
+
+		// build our msg
+		msg := h.Backend().NewIncomingMsg(c, urn, pmMsg.Content.Text).WithExternalID(pmMsg.ID)
+		msgs = append(msgs, msg)
+	}
+
+	// and finally write our message
+	return handlers.WriteMsgsAndResponse(ctx, h, msgs, w, r)
+}
+
+// SendMsg sends the passed in message, returning any error
+func (h *handler) SendMsg(_ context.Context, msg courier.Msg) (courier.MsgStatus, error) {
+	username := msg.Channel().StringConfigForKey(configUsername, "")
+	if username == "" {
+		return nil, fmt.Errorf("no username set for PM channel")
+	}
+
+	password := msg.Channel().StringConfigForKey(configPassword, "")
+	if password == "" {
+		return nil, fmt.Errorf("no password set for PM channel")
+	}
+
+	shortCode := msg.Channel().Address()
+	if shortCode == "" {
+		return nil, fmt.Errorf("no phone sender set for PM channel")
+	}
+
+	baseURL := msg.Channel().StringConfigForKey(configBaseURL, "")
+	if baseURL == "" {
+		return nil, fmt.Errorf("no base url set for PM channel")
+	}
+
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored)
+
+	for _, part := range handlers.SplitMsg(handlers.GetTextAndAttachments(msg), maxMsgLength) {
+		payload := mtPayload{}
+		message := mtMessage{}
+
+		message.Recipient = strings.TrimLeft(msg.URN().Path(), "+")
+		message.MessageID = msg.ID().String()
+		message.SMS.Originator = shortCode
+		message.SMS.Content.Text = part
+
+		payload.Messages = append(payload.Messages, message)
+
+		jsonBody, err := json.Marshal(payload)
+		if err != nil {
+			return status, err
```

Comment:
This shouldn't be returning an error, rather it should be setting the error on the status then returning that. (we should only return errors in send methods for things that are internal errors (ie, DB down, misconfiguration etc))

Reply:
Please, @nicpottier review the changes, thank you.

Reply:
@edudouglas I am so sorry but apparently having just come back from vacation I wasn't thinking clearly here, I though this was unmarshalling a response rather than building a request. Having `Marshal` fail does indeed fall into the category of things that are internal errors (it should never happen unless we run out of ram or something) so just returning the error (with a nil status) is indeed a valid thing to do here. Can you revert your change and modify this to just return `nil, err`?

Again, apologies for the mistake on my side.

Reply:
No problems @nicpottier. I reverted the code.